odoo/osv/expression.py

```diff
@@ -329,6 +329,40 @@ def distribute_not(domain):
 
     return result
 
+def remove_domain_leaf(domain, fields_to_remove, _normalized_domain=False):
+    """
+    each operator is a logic gate:
+    - '&' and '|' take two entries and can be ignored if one of them (or the two of them) has to be removed
+    -'!' take one entry and can be ignored if this entry has to be removed
+    - _normalized_domain should not be used and is used to know if the function is called by the user or not
+    return domain, rest_domain -> rest_domain should be empty if the operation is finished
+    """
+    _operator = ''
+    def _clean_result(result):
+        return result if _normalized_domain else result[0]
+    
+    if len(domain) == 0:
+        return _clean_result(([], []))
+    if not _normalized_domain: #First call of the function
+        domain = normalize_domain(domain)
```

Comment:
I am wondering if it will not be easier to directly distribute not operator. 🙂 
```suggestion
        domain = normalize_domain(domain)
        domain = distribute_not(domain)
```

However, without that, your solution seems to work as expected even with `!` 🤔 
I am not sure to understand when we can remove it. 🤔